File: modules/cubex-scheduler/src/main/java/org/cubexmc/scheduler/CubexScheduler.java
package org.cubexmc.scheduler;

import com.tcoded.folialib.FoliaLib;
import com.tcoded.folialib.wrapper.task.WrappedTask;
import java.util.Objects;
import java.util.Set;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ConcurrentHashMap;
import java.util.function.Consumer;
import org.bukkit.Bukkit;
import org.bukkit.Location;
import org.bukkit.entity.Entity;
import org.bukkit.plugin.Plugin;
import org.bukkit.scheduler.BukkitTask;
import org.cubexmc.core.CubexPlugin;

public final class CubexScheduler {

    private final Plugin plugin;
    private final Set<ManagedCubexTask> tasks = ConcurrentHashMap.newKeySet();
    private volatile FoliaLib foliaLib;
    private volatile boolean foliaInitFailed;

    private CubexScheduler(Plugin plugin) {
        this.plugin = Objects.requireNonNull(plugin, "plugin");
    }

    public static CubexScheduler bindTo(CubexPlugin plugin) {
        CubexScheduler scheduler = new CubexScheduler(plugin);
        plugin.bind((Runnable) scheduler::cancelAll);
        return scheduler;
    }

    public static CubexScheduler create(Plugin plugin) {
        if (plugin instanceof CubexPlugin) {
            return bindTo((CubexPlugin) plugin);
        }
        return new CubexScheduler(plugin);
    }

    public static boolean detectFolia() {
        try {
            Class.forName("io.papermc.paper.threadedregions.RegionizedServer");
            return true;
        } catch (ClassNotFoundException ignored) {
            return false;
        }
    }

    public Plugin plugin() {
        return plugin;
    }

    public boolean isFolia() {
        FoliaLib lib = foliaLib();
        return lib != null && lib.isFolia();
    }

    public boolean isPaper() {
        FoliaLib lib = foliaLib();
        return lib != null && lib.isPaper();
    }

    public boolean isSpigot() {
        FoliaLib lib = foliaLib();
        return lib != null && lib.isSpigot();
    }

    public CubexTask runGlobal(Runnable task) {
        return runGlobalLater(task, 0L);
    }

    public CubexTask runGlobal(Consumer<CubexTask> task) {
        return runGlobalLater(task, 0L);
    }

    public CubexTask runGlobalLater(Runnable task, long delayTicks) {
        return runGlobalLater(ignored -> task.run(), delayTicks);
    }

    public CubexTask runGlobalLater(Consumer<CubexTask> task, long delayTicks) {
        long delay = Math.max(0L, delayTicks);
        ManagedCubexTask managed = newTask();
        Runnable wrapped = oneShot(managed, task);
        FoliaLib lib = foliaLib();
        if (lib != null) {
            if (delay == 0L) {
                CompletableFuture<Void> future = lib.getScheduler().runNextTick(foliaTask -> {
                    managed.attach(foliaTask);
                    wrapped.run();
                });
                managed.attach(future);
            } else {
                WrappedTask nativeTask = lib.getScheduler().runLater(wrapped, delay);
                managed.attach(nativeTask);
            }
            return managed;
        }
        BukkitTask nativeTask = delay == 0L
                ? Bukkit.getScheduler().runTask(plugin, wrapped)
                : Bukkit.getScheduler().runTaskLater(plugin, wrapped, delay);
        managed.attach(nativeTask);
        return managed;
    }

    public CubexTask runGlobalTimer(Runnable task, long delayTicks, long periodTicks) {
        return runGlobalTimer(ignored -> task.run(), delayTicks, periodTicks);
    }

    public CubexTask runGlobalTimer(Consumer<CubexTask> task, long delayTicks, long periodTicks) {
        long delay = Math.max(1L, delayTicks);
        long period = Math.max(1L, periodTicks);
        ManagedCubexTask managed = newTask();
        Runnable wrapped = repeating(managed, task);
        FoliaLib lib = foliaLib();
        Object nativeTask = lib != null
                ? lib.getScheduler().runTimer(wrapped, delay, period)
                : Bukkit.getScheduler().runTaskTimer(plugin, wrapped, delay, period);
        managed.attach(nativeTask);
        return managed;
    }

    public CubexTask runAsync(Runnable task) {
        return runAsyncLater(task, 0L);
    }

    public CubexTask runAsync(Consumer<CubexTask> task) {
        return runAsyncLater(task, 0L);
    }

    public CubexTask runAsyncLater(Runnable task, long delayTicks) {
        return runAsyncLater(ignored -> task.run(), delayTicks);
    }

    public CubexTask runAsyncLater(Consumer<CubexTask> task, long delayTicks) {
        long delay = Math.max(0L, delayTicks);
        ManagedCubexTask managed = newTask();
        Runnable wrapped = oneShot(managed, task);
        FoliaLib lib = foliaLib();
        Object nativeTask = lib != null
                ? (delay == 0L ? lib.getScheduler().runAsync(foliaTask -> {
                    managed.attach(foliaTask);
                    wrapped.run();
                }) : lib.getScheduler().runLaterAsync(wrapped, delay))
                : Bukkit.getScheduler().runTaskLaterAsynchronously(plugin, wrapped, delay);
        managed.attach(nativeTask);
        return managed;
    }

    public CubexTask runAsyncTimer(Runnable task, long delayTicks, long periodTicks) {
        return runAsyncTimer(ignored -> task.run(), delayTicks, periodTicks);
    }

    public CubexTask runAsyncTimer(Consumer<CubexTask> task, long delayTicks, long periodTicks) {
        long delay = Math.max(1L, delayTicks);
        long period = Math.max(1L, periodTicks);
        ManagedCubexTask managed = newTask();
        Runnable wrapped = repeating(managed, task);
        FoliaLib lib = foliaLib();
        Object nativeTask = lib != null
                ? lib.getScheduler().runTimerAsync(wrapped, delay, period)
                : Bukkit.getScheduler().runTaskTimerAsynchronously(plugin, wrapped, delay, period);
        managed.attach(nativeTask);
        return managed;
    }

    public CubexTask runAtEntity(Entity entity, Runnable task) {
        return runAtEntityLater(entity, task, 0L);
    }

    public CubexTask runAtEntity(Entity entity, Consumer<CubexTask> task) {
        return runAtEntityLater(entity, task, 0L);
    }

    public CubexTask runAtEntityLater(Entity entity, Runnable task, long delayTicks) {
        return runAtEntityLater(entity, ignored -> task.run(), delayTicks);
    }

    public CubexTask runAtEntityLater(Entity entity, Consumer<CubexTask> task, long delayTicks) {
        Objects.requireNonNull(entity, "entity");
        long delay = Math.max(0L, delayTicks);
        ManagedCubexTask managed = newTask();
        Runnable wrapped = oneShot(managed, task);
        FoliaLib lib = foliaLib();
        if (lib != null) {
            if (delay == 0L) {
                CompletableFuture<?> future = lib.getScheduler().runAtEntity(entity, foliaTask -> {
                    managed.attach(foliaTask);
                    wrapped.run();
                });
                managed.attach(future);
            } else {
                managed.attach(lib.getScheduler().runAtEntityLater(entity, wrapped, delay));
            }
            return managed;
        }
        BukkitTask nativeTask = delay == 0L
                ? Bukkit.getScheduler().runTask(plugin, wrapped)
                : Bukkit.getScheduler().runTaskLater(plugin, wrapped, delay);
        managed.attach(nativeTask);
        return managed;
    }

    public CubexTask runAtEntityTimer(Entity entity, Runnable task, long delayTicks, long periodTicks) {
        return runAtEntityTimer(entity, ignored -> task.run(), delayTicks, periodTicks);
    }

    public CubexTask runAtEntityTimer(Entity entity, Consumer<CubexTask> task, long delayTicks, long periodTicks) {
        Objects.requireNonNull(entity, "entity");
        long delay = Math.max(1L, delayTicks);
        long period = Math.max(1L, periodTicks);
        ManagedCubexTask managed = newTask();
        Runnable wrapped = repeating(managed, task);
        FoliaLib lib = foliaLib();
        Object nativeTask = lib != null
                ? lib.getScheduler().runAtEntityTimer(entity, wrapped, delay, period)
                : Bukkit.getScheduler().runTaskTimer(plugin, wrapped, delay, period);
        managed.attach(nativeTask);
        return managed;
    }

    public CubexTask runAtLocation(Location location, Runnable task) {
        return runAtLocationLater(location, task, 0L);
    }

    public CubexTask runAtLocation(Location location, Consumer<CubexTask> task) {
        return runAtLocationLater(location, task, 0L);
    }

    public CubexTask runAtLocationLater(Location location, Runnable task, long delayTicks) {
        return runAtLocationLater(location, ignored -> task.run(), delayTicks);
    }

    public CubexTask runAtLocationLater(Location location, Consumer<CubexTask> task, long delayTicks) {
        Objects.requireNonNull(location, "location");
        long delay = Math.max(0L, delayTicks);
        ManagedCubexTask managed = newTask();
        Runnable wrapped = oneShot(managed, task);
        FoliaLib lib = foliaLib();
        if (lib != null) {
            if (delay == 0L) {
                CompletableFuture<Void> future = lib.getScheduler().runAtLocation(location, foliaTask -> {
                    managed.attach(foliaTask);
                    wrapped.run();
                });
                managed.attach(future);
            } else {
                managed.attach(lib.getScheduler().runAtLocationLater(location, wrapped, delay));
            }
            return managed;
        }
        BukkitTask nativeTask = delay == 0L
                ? Bukkit.getScheduler().runTask(plugin, wrapped)
                : Bukkit.getScheduler().runTaskLater(plugin, wrapped, delay);
        managed.attach(nativeTask);
        return managed;
    }

    public CubexTask runAtLocationTimer(Location location, Runnable task, long delayTicks, long periodTicks) {
        return runAtLocationTimer(location, ignored -> task.run(), delayTicks, periodTicks);
    }

    public CubexTask runAtLocationTimer(Location location, Consumer<CubexTask> task, long delayTicks, long periodTicks) {
        Objects.requireNonNull(location, "location");
        long delay = Math.max(1L, delayTicks);
        long period = Math.max(1L, periodTicks);
        ManagedCubexTask managed = newTask();
        Runnable wrapped = repeating(managed, task);
        FoliaLib lib = foliaLib();
        Object nativeTask = lib != null
                ? lib.getScheduler().runAtLocationTimer(location, wrapped, delay, period)
                : Bukkit.getScheduler().runTaskTimer(plugin, wrapped, delay, period);
        managed.attach(nativeTask);
        return managed;
    }

    public CompletableFuture<Boolean> teleportAsync(Entity entity, Location location) {
        Objects.requireNonNull(entity, "entity");
        Objects.requireNonNull(location, "location");
        FoliaLib lib = foliaLib();
        if (lib != null) {
            return lib.getScheduler().teleportAsync(entity, location);
        }
        try {
            return CompletableFuture.completedFuture(entity.teleport(location));
        } catch (Throwable throwable) {
            CompletableFuture<Boolean> failed = new CompletableFuture<>();
            failed.completeExceptionally(throwable);
            return failed;
        }
    }

    public void cancelAll() {
        for (ManagedCubexTask task : tasks.toArray(new ManagedCubexTask[0])) {
            task.cancel();
        }
        FoliaLib lib = foliaLib;
        if (lib != null) {
            try {
                lib.getScheduler().cancelAllTasks();
            } catch (Throwable ignored) {
            }
        }
    }

    private ManagedCubexTask newTask() {
        final ManagedCubexTask[] holder = new ManagedCubexTask[1];
        holder[0] = new ManagedCubexTask(() -> tasks.remove(holder[0]));
        tasks.add(holder[0]);
        return holder[0];
    }

    private Runnable oneShot(ManagedCubexTask managed, Consumer<CubexTask> task) {
        Objects.requireNonNull(task, "task");
        return () -> {
            if (managed.isCancelled()) {
                return;
            }
            try {
                task.accept(managed);
            } finally {
                managed.complete();
            }
        };
    }

    private Runnable repeating(ManagedCubexTask managed, Consumer<CubexTask> task) {
        Objects.requireNonNull(task, "task");
        return () -> {
            if (!managed.isCancelled()) {
                task.accept(managed);
            }
        };
    }

    private FoliaLib foliaLib() {
        if (foliaInitFailed) {
            return null;
        }
        FoliaLib lib = foliaLib;
        if (lib != null) {
            return lib;
        }
        try {
            lib = new FoliaLib(plugin);
            foliaLib = lib;
            return lib;
        } catch (Throwable ignored) {
            foliaInitFailed = true;
            return null;
        }
    }
}

File: modules/cubex-scheduler/src/main/kotlin/org/cubexmc/scheduler/CubexScheduler.kt
package org.cubexmc.scheduler

import com.tcoded.folialib.FoliaLib
import org.bukkit.Bukkit
import org.bukkit.Location
import org.bukkit.entity.Entity
import org.bukkit.plugin.Plugin
import org.cubexmc.core.CubexPlugin
import java.util.concurrent.CompletableFuture
import java.util.concurrent.ConcurrentHashMap
import java.util.function.Consumer
import kotlin.math.max

class CubexScheduler private constructor(private val plugin: Plugin) {
    private val tasks: MutableSet<ManagedCubexTask> = ConcurrentHashMap.newKeySet()

    @Volatile
    private var foliaLib: FoliaLib? = null

    @Volatile
    private var foliaInitFailed = false

    fun plugin(): Plugin = plugin

    val isFolia: Boolean
        get() = foliaLib()?.isFolia == true

    val isPaper: Boolean
        get() = foliaLib()?.isPaper == true

    val isSpigot: Boolean
        get() = foliaLib()?.isSpigot == true

    fun runGlobal(task: Runnable): CubexTask = runGlobalLater(task, 0L)

    fun runGlobal(task: Consumer<CubexTask>): CubexTask = runGlobalLater(task, 0L)

    fun runGlobalLater(task: Runnable, delayTicks: Long): CubexTask =
        runGlobalLater(Consumer { task.run() }, delayTicks)

    fun runGlobalLater(task: Consumer<CubexTask>, delayTicks: Long): CubexTask {
        val delay = max(0L, delayTicks)
        val managed = newTask()
        val wrapped = oneShot(managed, task)
        val lib = foliaLib()
        if (lib != null) {
            if (delay == 0L) {
                val future = lib.scheduler.runNextTick { foliaTask ->
                    managed.attach(foliaTask)
                    wrapped.run()
                }
                managed.attach(future)
            } else {
                managed.attach(lib.scheduler.runLater(wrapped, delay))
            }
            return managed
        }
        managed.attach(
            if (delay == 0L) Bukkit.getScheduler().runTask(plugin, wrapped)
            else Bukkit.getScheduler().runTaskLater(plugin, wrapped, delay),
        )
        return managed
    }

    fun runGlobalTimer(task: Runnable, delayTicks: Long, periodTicks: Long): CubexTask =
        runGlobalTimer(Consumer { task.run() }, delayTicks, periodTicks)

    fun runGlobalTimer(task: Consumer<CubexTask>, delayTicks: Long, periodTicks: Long): CubexTask {
        val delay = max(1L, delayTicks)
        val period = max(1L, periodTicks)
        val managed = newTask()
        val wrapped = repeating(managed, task)
        managed.attach(
            foliaLib()?.scheduler?.runTimer(wrapped, delay, period)
                ?: Bukkit.getScheduler().runTaskTimer(plugin, wrapped, delay, period),
        )
        return managed
    }

    fun runAsync(task: Runnable): CubexTask = runAsyncLater(task, 0L)

    fun runAsync(task: Consumer<CubexTask>): CubexTask = runAsyncLater(task, 0L)

    fun runAsyncLater(task: Runnable, delayTicks: Long): CubexTask =
        runAsyncLater(Consumer { task.run() }, delayTicks)

    fun runAsyncLater(task: Consumer<CubexTask>, delayTicks: Long): CubexTask {
        val delay = max(0L, delayTicks)
        val managed = newTask()
        val wrapped = oneShot(managed, task)
        val lib = foliaLib()
        managed.attach(
            if (lib == null) {
                Bukkit.getScheduler().runTaskLaterAsynchronously(plugin, wrapped, delay)
            } else if (delay == 0L) {
                lib.scheduler.runAsync { foliaTask ->
                    managed.attach(foliaTask)
                    wrapped.run()
                }
            } else {
                lib.scheduler.runLaterAsync(wrapped, delay)
            },
        )
        return managed
    }

    fun runAsyncTimer(task: Runnable, delayTicks: Long, periodTicks: Long): CubexTask =
        runAsyncTimer(Consumer { task.run() }, delayTicks, periodTicks)

    fun runAsyncTimer(task: Consumer<CubexTask>, delayTicks: Long, periodTicks: Long): CubexTask {
        val delay = max(1L, delayTicks)
        val period = max(1L, periodTicks)
        val managed = newTask()
        val wrapped = repeating(managed, task)
        managed.attach(
            foliaLib()?.scheduler?.runTimerAsync(wrapped, delay, period)
                ?: Bukkit.getScheduler().runTaskTimerAsynchronously(plugin, wrapped, delay, period),
        )
        return managed
    }

    fun runAtEntity(entity: Entity, task: Runnable): CubexTask = runAtEntityLater(entity, task, 0L)

    fun runAtEntity(entity: Entity, task: Consumer<CubexTask>): CubexTask = runAtEntityLater(entity, task, 0L)

    fun runAtEntityLater(entity: Entity, task: Runnable, delayTicks: Long): CubexTask =
        runAtEntityLater(entity, Consumer { task.run() }, delayTicks)

    fun runAtEntityLater(entity: Entity, task: Consumer<CubexTask>, delayTicks: Long): CubexTask {
        val delay = max(0L, delayTicks)
        val managed = newTask()
        val wrapped = oneShot(managed, task)
        val lib = foliaLib()
        if (lib != null) {
            managed.attach(
                if (delay == 0L) {
                    lib.scheduler.runAtEntity(entity) { foliaTask ->
                        managed.attach(foliaTask)
                        wrapped.run()
                    }
                } else {
                    lib.scheduler.runAtEntityLater(entity, wrapped, delay)
                },
            )
            return managed
        }
        managed.attach(
            if (delay == 0L) Bukkit.getScheduler().runTask(plugin, wrapped)
            else Bukkit.getScheduler().runTaskLater(plugin, wrapped, delay),
        )
        return managed
    }

    fun runAtEntityTimer(entity: Entity, task: Runnable, delayTicks: Long, periodTicks: Long): CubexTask =
        runAtEntityTimer(entity, Consumer { task.run() }, delayTicks, periodTicks)

    fun runAtEntityTimer(
        entity: Entity,
        task: Consumer<CubexTask>,
        delayTicks: Long,
        periodTicks: Long,
    ): CubexTask {
        val delay = max(1L, delayTicks)
        val period = max(1L, periodTicks)
        val managed = newTask()
        val wrapped = repeating(managed, task)
        managed.attach(
            foliaLib()?.scheduler?.runAtEntityTimer(entity, wrapped, delay, period)
                ?: Bukkit.getScheduler().runTaskTimer(plugin, wrapped, delay, period),
        )
        return managed
    }

    fun runAtLocation(location: Location, task: Runnable): CubexTask = runAtLocationLater(location, task, 0L)

    fun runAtLocation(location: Location, task: Consumer<CubexTask>): CubexTask = runAtLocationLater(location, task, 0L)

    fun runAtLocationLater(location: Location, task: Runnable, delayTicks: Long): CubexTask =
        runAtLocationLater(location, Consumer { task.run() }, delayTicks)

    fun runAtLocationLater(location: Location, task: Consumer<CubexTask>, delayTicks: Long): CubexTask {
        val delay = max(0L, delayTicks)
        val managed = newTask()
        val wrapped = oneShot(managed, task)
        val lib = foliaLib()
        if (lib != null) {
            managed.attach(
                if (delay == 0L) {
                    lib.scheduler.runAtLocation(location) { foliaTask ->
                        managed.attach(foliaTask)
                        wrapped.run()
                    }
                } else {
                    lib.scheduler.runAtLocationLater(location, wrapped, delay)
                },
            )
            return managed
        }
        managed.attach(
            if (delay == 0L) Bukkit.getScheduler().runTask(plugin, wrapped)
            else Bukkit.getScheduler().runTaskLater(plugin, wrapped, delay),
        )
        return managed
    }

    fun runAtLocationTimer(location: Location, task: Runnable, delayTicks: Long, periodTicks: Long): CubexTask =
        runAtLocationTimer(location, Consumer { task.run() }, delayTicks, periodTicks)

    fun runAtLocationTimer(
        location: Location,
        task: Consumer<CubexTask>,
        delayTicks: Long,
        periodTicks: Long,
    ): CubexTask {
        val delay = max(1L, delayTicks)
        val period = max(1L, periodTicks)
        val managed = newTask()
        val wrapped = repeating(managed, task)
        managed.attach(
            foliaLib()?.scheduler?.runAtLocationTimer(location, wrapped, delay, period)
                ?: Bukkit.getScheduler().runTaskTimer(plugin, wrapped, delay, period),
        )
        return managed
    }

    fun teleportAsync(entity: Entity, location: Location): CompletableFuture<Boolean> {
        val lib = foliaLib()
        if (lib != null) return lib.scheduler.teleportAsync(entity, location)
        return try {
            CompletableFuture.completedFuture(entity.teleport(location))
        } catch (throwable: Throwable) {
            CompletableFuture<Boolean>().also { it.completeExceptionally(throwable) }
        }
    }

    fun cancelAll() {
        tasks.toTypedArray().forEach { it.cancel() }
        foliaLib?.let { lib ->
            try {
                lib.scheduler.cancelAllTasks()
            } catch (_: Throwable) {
                // Best-effort cleanup during shutdown.
            }
        }
    }

    private fun newTask(): ManagedCubexTask {
        lateinit var managed: ManagedCubexTask
        managed = ManagedCubexTask { tasks.remove(managed) }
        tasks.add(managed)
        return managed
    }

    private fun oneShot(managed: ManagedCubexTask, task: Consumer<CubexTask>): Runnable = Runnable {
        if (managed.isCancelled()) return@Runnable
        try {
            task.accept(managed)
        } finally {
            managed.complete()
        }
    }

    private fun repeating(managed: ManagedCubexTask, task: Consumer<CubexTask>): Runnable = Runnable {
        if (!managed.isCancelled()) task.accept(managed)
    }

    private fun foliaLib(): FoliaLib? {
        if (foliaInitFailed) return null
        foliaLib?.let { return it }
        return try {
            FoliaLib(plugin).also { foliaLib = it }
        } catch (_: Throwable) {
            foliaInitFailed = true
            null
        }
    }

    companion object {
        @JvmStatic
        fun bindTo(plugin: CubexPlugin): CubexScheduler =
            CubexScheduler(plugin).also { scheduler -> plugin.bind(Runnable { scheduler.cancelAll() }) }

        @JvmStatic
        fun create(plugin: Plugin): CubexScheduler =
            if (plugin is CubexPlugin) bindTo(plugin) else CubexScheduler(plugin)

        @JvmStatic
        fun detectFolia(): Boolean = try {
            Class.forName("io.papermc.paper.threadedregions.RegionizedServer")
            true
        } catch (_: ClassNotFoundException) {
            false
        }
    }
}
